core/lib/server/pub.ts

```diff
@@ -589,3 +600,551 @@ export const deletePub = async (pubId: PubsId) =>
+	 * Only used for testing.
+	 * If true the raw result of the query is returned, without nesting the values and children.
+	 */
+	_debugDontNest?: boolean;
+	fieldSlugs?: string[];
+} & GetManyParams;
+
+type PubIdOrPubTypeIdOrStageIdOrCommunityId =
+	| {
+			pubId: PubsId;
+			pubTypeId?: never;
+			stageId?: never;
+			communityId?: never;
+	  }
+	| {
+			pubId?: never;
+			pubTypeId: PubTypesId;
+			stageId?: never;
+			communityId?: never;
+	  }
+	| {
+			pubId?: never;
+			pubTypeId?: never;
+			stageId: StagesId;
+			communityId?: never;
+	  }
+	| {
+			pubId?: never;
+			pubTypeId?: never;
+			stageId?: never;
+			communityId: CommunitiesId;
+	  };
+
+/**
+ * Only add the `children` if the `withChildren` option has not been set to `false
+ */
+type MaybePubChildren<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withChildren"] extends false
+		? { children?: never }
+		: { children: ProcessedPub<Options>[] };
+
+/**
+ * Only add the `stage` if the `withStage` option has not been set to `false
+ */
+type MaybePubStage<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withStage"] extends true ? { stage: Stages } : { stage?: never };
+
+/**
+ * Only add the `pubType` if the `withPubType` option has not been set to `false
+ */
+type MaybePubPubType<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withPubType"] extends true ? { pubType: PubTypes } : { pubType?: never };
+
+type MaybeWithRelatedPub<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withRelatedPubs"] extends false
+		? { relatedPub?: never }
+		: { relatedPub: ProcessedPub<Options>[] };
+
+/**
+ * Those options of `GetPubsWithRelatedValuesAndChildrenOptions` that affect the output of `ProcessedPub`
+ *
+ * This way it's more easy to specify what kind of `ProcessedPub` we want as e.g. the input type of a function
+ *
+ **/
+type MaybeOptions = Pick<
+	GetPubsWithRelatedValuesAndChildrenOptions,
+	"withChildren" | "withRelatedPubs" | "withPubType" | "withStage"
+>;
+
+export type ProcessedPub<Options extends MaybeOptions = {}> = {
+	id: PubsId;
+	stageId: StagesId | null;
+	communityId: CommunitiesId;
+	pubTypeId: PubTypesId;
+	parentId: PubsId | null;
+	/**
+	 * An array of values for the pub.
+	 */
+	values: ({
+		id: PubValuesId;
+		fieldId: PubFieldsId;
+		value: unknown;
+		relatedPubId: PubsId | null;
+		createdAt: Date;
+		updatedAt: Date;
+		/**
+		 * Information about the field that the value belongs to.
+		 */
+		schemaName: CoreSchemaType;
+		fieldSlug: string;
+	} & MaybeWithRelatedPub<Options>)[];
+	createdAt: Date;
+	/**
+	 * The `updatedAt` of the latest value, or of the pub if the pub itself has a higher `updatedAt` or if there are no values
+	 *
+	 * We do this because the Pub itself is rarely if ever changed over time.
+	 * TODO: Possibly add the `updatedAt` of `PubsInStages` here as well?
+	 * At time of writing (2024/11/04) I don't think that table has an `updatedAt`.
```

Comment:
It seems like we have two options when dealing with a pub's updatedAt: either look at related tables (values) to calculate the user-facing updatedAt during read queries, or update it during write queries that only modify values. I think this is a more general problem with the application, which also affects forms right now, and probably will affect just about anything where we want to show a last modified date!

IMO we should agree on one of these two strategies for simplicity. I'm slightly in favor of doing it on writes rather than reads, because I think those queries will wind up simpler in the long run. So instead of using the most recent pub value's updatedAt here (or even the last time the pub was moved to a new stage) we would instead explicitly update a pub's updatedAt each time we moved it to a new stage or updated its values. But tbh I don't feel that strongly or really want you to change this now, I'd just like to make sure we agree on a strategy!

Reply:
hmm yeah i agree, that makes more sense. There'll probably some queries where we fetch pubs where we don't want to fetch the values, and needing that to calculate the updatedAt is a lot of extra overhead, plus lowers the cache hit rates. I won't change it now, but we will have to do that in the future.